### PR TITLE
Use enums to represent the clients (#1395)

### LINKED ISSUE #1395: Use Enums to represent Index and Object storage types
See: https://github.com/cortexproject/cortex/pull/1390

We have to manually update the error to have all the options, instead we could just generate those using enums.

### PR DESCRIPTION
**Pull request by ChinYing-Li**
Signed-off-by: ChinYing-Li <chinying.li@mail.utoronto.ca>

**What this PR does**:
Use enums (type string) to represent the name of clients.
If this is what @gouthamve meant by using "enums", should we also replace the case conditions with enums? e.g. 
`case "aws":` to `case AWS:`?

**Which issue(s) this PR fixes**:
Fixes #1395

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@
 * [ENHANCEMENT] Distributor: reduce CPU and memory when an high number of errors are returned by the distributor on the write path. #3990
 * [ENHANCEMENT] Put metric before label value in the "label value too long" error message. #4018
 * [ENHANCEMENT] Allow use of `y|w|d` suffixes for duration related limits and per-tenant limits. #4044
+* [ENHANCEMENT] Use enums to represent client types in factory functions. #4048
```
Comment (jtlisi) [inline CHANGELOG.md:27]: **nit:** this change shouldn't have any behavioral changes for the user so there is no need for a changelog entry.

```suggestion
```

**Review by jtlisi — CHANGES_REQUESTED** [review]
LGTM, thanks for taking care of this. Added one minor nit.

File: pkg/chunk/storage/factory.go [hunk 2]
```diff
@@ -32,6 +32,21 @@ const (
...
+	Azure          = "azure"
+	Cassandra      = "cassandra"
+	InMemory       = "inmemory"
+	BigTable       = "bigtable"
+	BigTableHashed = "bigtable-hashed"
+	FileSystem     = "filesystem"
+	GCP            = "gcp"
+	GCS            = "gcs"
+	GrpcStore      = "grpc-store"
+	S3             = "s3"
```
Comment (jtlisi) [inline pkg/chunk/storage/factory.go:47]: **suggestion(optional):** for enums like this, I think it makes sense to prefix them with some kind of consistent identifier. Out of context (like in an editor autocomplete dropdown) these variable names could be confusing but with a consistent prefix you know they are related. The other option is to make a custom enum type on top of the string type, [this site shows an example using an int](https://www.educative.io/edpresso/what-is-an-enum-in-golang).

```suggestion
	StoreTypeAWS            = "aws"
	StoreTypeAzure          = "azure"
	StoreTypeCassandra      = "cassandra"
	StoreTypeInMemory       = "inmemory"
	StoreTypeBigTable       = "bigtable"
	StoreTypeBigTableHashed = "bigtable-hashed"
	StoreTypeFileSystem     = "filesystem"
	StoreTypeGCP            = "gcp"
	StoreTypeGCS            = "gcs"
	StoreTypeGrpc      = "grpc-store"
	StoreTypeS3             = "s3"
```

Comment (ChinYing-Li) [reply pkg/chunk/storage/factory.go]: I added the prefix `StorageType` to follow the style here: https://github.com/cortexproject/cortex/blob/cb8ed0e4eaaa73f91881a13e57c014bfffa468a2/pkg/chunk/storage/factory.go#L30-L33

**Review by jtlisi — APPROVED** [review]
LGTM

**Review by pracucci — APPROVED** [review]
Thanks!

**Pull request merged**